### ISSUE TITLE
AWS EKS OIDC - could not get secret data from provider

### ISSUE THREAD

**Issue by shllg** [external-secrets/external-secrets#1555]
**Issue**

Although controller authenticates via ServiceAccount / OIDC, fetching secrets always ends up in "could not get secret data from provider" error message. 

```
{"level":"error","ts":1663325349.8799698,"logger":"controllers.ExternalSecret","msg":"could not get secret data from provider","ExternalSecret":"[NAMESPACE]/[SECRET-NAME]","SecretStore":"/aws-secret-store","error":"RequestError: send request failed\ncaused by: Post \"https://secretsmanager.eu-central-1.amazonaws.com/\": dial tcp XXX.XXX.XXX.XXX:443: i/o timeout","stacktrace":"sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Reconcile\n\t/home/runner/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.12.3/pkg/internal/controller/controller.go:121\nsigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).reconcileHandler\n\t/home/runner/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.12.3/pkg/internal/controller/controller.go:320\nsigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).processNextWorkItem\n\t/home/runner/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.12.3/pkg/internal/controller/controller.go:273\nsigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Start.func2.2\n\t/home/runner/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.12.3/pkg/internal/controller/controller.go:234"}
```

**Setup**

EKS Cluster with connected OIDC provider is configured.

*IAM Policy*

```json
{
    "Version": "2012-10-17",
    "Statement": [
        {
            "Effect": "Allow",
            "Action": [
                "secretsmanager:GetResourcePolicy",
                "secretsmanager:GetSecretValue",
                "secretsmanager:DescribeSecret",
                "secretsmanager:ListSecretVersionIds"
            ],
            "Resource": "arn:aws:secretsmanager:eu-central-1:[ACCOUNT-ID]:secret:some-prefix/*"
        }
    ]
}
```

*IAM Role*

Policy is assigned to an IAM Role called below `IAM_ROLE_ARN`.

Trust relationships of `IAM_ROLE_ARN`:

```json
{
    "Version": "2012-10-17",
    "Statement": [
        {
            "Effect": "Allow",
            "Principal": {
                "Federated": "arn:aws:iam::[ACCOUNT-ID]:oidc-provider/oidc.eks.eu-central-1.amazonaws.com/id/[OIDC-PROVIDER-ID]"
            },
            "Action": "sts:AssumeRoleWithWebIdentity",
            "Condition": {
                "StringLike": {
                    "oidc.eks.eu-central-1.amazonaws.com/id/[OIDC-PROVIDER-ID]:sub": "system:serviceaccount:external-dns:external-dns",
                    "oidc.eks.eu-central-1.amazonaws.com/id/[OIDC-PROVIDER-ID]:aud": "sts.amazonaws.com"
                }
            }
        }
    ]
}
```

*K8s ServiceAccount* 

```yaml
apiVersion: v1
kind: ServiceAccount
metadata:
  name: external-secrets
  namespace: external-secrets
  annotations:
    eks.amazonaws.com/role-arn: [IAM_ROLE_ARN]
```

*K8s ClusterSecretStore*

```yaml
apiVersion: external-secrets.io/v1beta1
kind: ClusterSecretStore
metadata:
  name: aws-secret-store
  namespace: external-secrets
spec:
  provider:
    aws:
      service: SecretsManager
      region: ${AWS_REGION}
      auth:
        jwt:
          serviceAccountRef:
            name: external-secrets
            namespace: external-secrets
```

*K8s Test Secret* 

```yaml
apiVersion: external-secrets.io/v1beta1
kind: ExternalSecret
metadata:
  name: external-secret-test
  namespace: hello-kubernetes
spec:
  refreshInterval: 1h
  secretStoreRef:
    name: aws-secret-store
    kind: ClusterSecretStore
  target:
    name: test-secret
    creationPolicy: Owner
  dataFrom:
  - extract:
      key: some-prefix/test-secret
```

---

**Observations**

- The OIDC token is issued which means the role assumption works. Changing something inside the IAM Role trust relationship which is not correct leads to an STS assumption error - as cross check.
- Changing the policy to `"Resource": "*"` does not help (do we need to fetch a new token on policy change?)
- Changing the trust relationships to `"oidc.eks.eu-central-1.amazonaws.com/id/[OIDC-PROVIDER-ID]:sub": "system:serviceaccount:*:*"` does not help (again I think that it's not an OIDC assumption issue)
- Security Groups allow all outbound traffic

---

Can somebody help me out with that issue? I don't know how to debug further. The setup looks fine up to the actual secret fetching. 

Thanks in advance :-)

**Comment by moolen** [comment]
This is a networking issue:
```
 Post \"https://secretsmanager.eu-central-1.amazonaws.com/\": dial tcp XXX.XXX.XXX.XXX:443: i/o timeout
```
Can you check egress traffic: EC2 security groups, NACL, NetworkPolicies and the ip address of :point_up_2: is it public or private? in case it's a private endpoint you must verify ingress there, too.

**Comment by shllg** [comment]
@moolen Thank you for the very fast response!

**All Security groups have the same "Outbound rules" setting for now:**

<img width="1072" alt="Bildschirmfoto 2022-09-16 um 13 56 18" src="https://user-images.githubusercontent.com/5073881/190633688-5db9df96-484b-461b-ac0c-7666e26cd133.png">

**VPC Network ACL has it's default settings (which should allow all egress traffic too) like:**

<img width="859" alt="Bildschirmfoto 2022-09-16 um 13 57 56" src="https://user-images.githubusercontent.com/5073881/190633839-28410c87-d3ab-4f9d-b192-e056c94bb887.png">

**Firewall policies:**

*none*

**VPC Route tables:**

- private subnets route to NAT
- public like that:

<img width="909" alt="Bildschirmfoto 2022-09-16 um 14 00 01" src="https://user-images.githubusercontent.com/5073881/190634197-d8755b48-6f82-4b89-918e-7fb08bd39296.png">

**IP address (like asked above):**

Always 10.0.X.X addresses. The external-secrets pods are running in a private subnet. I'll test it now to move them to public subnet. Still it should work in the private subnet due to the NAT entry in the routing table, right?

**Private Endpoint ingress**

> in case it's a private endpoint you must verify ingress there, too.

Can you help me out with that? Why is the ingress relevant here?

**Comment by moolen** [comment]
> in case it's a private endpoint you must verify ingress there, too.

This: `secretsmanager.eu-central-1.amazonaws.com` can point to a VPC-internal address ([like described here](https://docs.aws.amazon.com/secretsmanager/latest/userguide/vpc-endpoint-overview.html)). If that is the case you must verify the ingress rule on that particular endpoint (not on the external-secrets pod).

> private subnet + NAT

This works fine, that shouldn't be an issue!


All of that stuff looks good so far.
If you can SSH into the node (or start a pod with hostNetwork=true) i'd first check the connection Node -> Secretsmanager. If that works the issue is within Kubernetes: Either a NetworkPolicy or CNI-specific rules apply (like CiliumNetworkPolicy/ClusterCiliumNetworkPolicy or the same with calico).

**Comment by shllg** [comment]
🥳  I forgot that I created an endpoint to the secrets manager inside my RDS CloudFormation template. It is applied to the same private subnets that are used by the EKS Cluster (I guess that's not a good pattern...). Adding the node security group to that endpoint made it work. 

But I needed to move the pods back into the private subnets.

As I understood, it is not necessary to create endpoints, right? If the pods are running in the public subnets (which are not assigned to any endpoint), it should have worked correctly. Instead I'm getting STS connection errors here (`WebIdentityErr: failed to retrieve credentials\ncaused by: RequestError: send request failed\ncaused by: Post \"https://sts.eu-central-1.amazonaws.com/\": dial tcp 54.239.54.207:443`).

Well I think that part is offtopic here and the issue itself is solved 👍 .

---

@moolen We're paying for a support plan and didn't get - unfortunately - any helpful answer for a while. I'm very grateful you helped me out here... so fast...
